server/src/main/java/org/candlepin/model/activationkeys/ActivationKey.java

```diff
@@ -189,6 +186,71 @@ public void setOwner(Owner owner) {
         this.owner = owner;
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return String.format("ActivationKey [id: %s, name: %s, description: %s]",
+                this.getId(), this.getName(), this.getDescription());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        }
+
+        if (obj instanceof ActivationKey && super.equals(obj)) {
+            ActivationKey that = (ActivationKey) obj;
+
+            // Pull the owner IDs, as we're not interested in verifying that the owners
+            // themselves are equal; just so long as they point to the same owner.
+            String thisOwnerId = this.getOwner() != null ? this.getOwner().getId() : null;
+            String thatOwnerId = that.getOwner() != null ? that.getOwner().getId() : null;
+
+            EqualsBuilder builder = new EqualsBuilder()
+                .append(this.getId(), that.getId())
+                .append(this.getName(), that.getName())
+                .append(this.getDescription(), that.getDescription())
+                .append(thisOwnerId, thatOwnerId)
+                .append(this.getReleaseVer(), that.getReleaseVer())
+                .append(this.getServiceLevel(), that.getServiceLevel())
+                .append(this.isAutoAttach(), that.isAutoAttach())
+                .append(this.getPools(), that.getPools())
+                .append(this.getProducts(), that.getProducts());
+
+            return builder.isEquals();
+        }
+
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        // Like with the equals method, we are not interested in hashing nested objects; we're only
+        // concerned with the reference to such an object.
+        HashCodeBuilder builder = new HashCodeBuilder(37, 7)
+            .append(super.hashCode())
+            .append(this.getId())
+            .append(this.getName())
+            .append(this.getDescription())
+            .append(this.getOwner() != null ? this.getOwner().getId() : null)
+            .append(this.getReleaseVer())
+            .append(this.getServiceLevel())
+            .append(this.isAutoAttach())
+            .append(this.getPools())
+            .append(this.getProducts());
```

Comment:
Same deal here. Hibernate's containers tend to not adhere to the contract on .equals and .hashCode, so make sure you're not at risk of running into the issues such violations present. Product.hashCode has another example for this scenario.